File: 90-subsets-ii/subsets-ii.cpp
class Solution {
public:
    vector<vector<int>> arr;
    void helper(vector<int>& nums,vector<int>& v,int ind,int size) {
        if(ind >= size) {
            arr.push_back(v); 
            return ;
        }
        v.push_back(nums[ind]);
        helper(nums,v,ind + 1,size);
        v.pop_back();
        int i = ind + 1;
        while(i < nums.size() && nums[i] == nums[ i - 1]) i++;
        helper(nums,v,i,size);
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        int n = nums.size();
        vector<int> v;
        sort(begin(nums),end(nums));
        helper(nums,v,0,n);
        return arr;
    }
};

File: 90-subsets-ii/subsets-ii.java

class Solution {
    List<List<Integer>> al = new ArrayList<>();
    HashMap<List<Integer>, Integer> hm = new HashMap<>();
    public void helper(int[] nums, List<Integer> list, int idx, int n) {
        if (idx >= n) {
            List<Integer> temp = new ArrayList<>(list);
            if (!hm.containsKey(temp)) {
                al.add(temp);
                hm.put(temp, hm.getOrDefault(temp, 0) + 1);
            }
            return;
        }
        list.add(nums[idx]);
        helper(nums, list, idx + 1, n);
        list.remove(list.size() - 1); 
        helper(nums, list, idx + 1, n);
    }
    
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums); 
        List<Integer> list = new ArrayList<>();
        helper(nums, list, 0, nums.length);
        return al;
    }
}
